spectator/stateless_meters.h

```diff
@@ -6,14 +6,39 @@
 namespace spectator {
 
 namespace detail {
+
+#include "valid_chars.inc"
+
 inline std::string as_string(std::string_view v) { 
     return {v.data(), v.size()}; 
 }
+
+inline bool contains_non_atlas_char(const std::string& input) {
+  return std::any_of(input.begin(), input.end(), [](char c) { return !kAtlasChars[c]; });
+}
+
+inline std::string replace_invalid_characters(const std::string& input) {
+  if (contains_non_atlas_char(input)) {
+    std::string result{input};
+    for (char &c : result) {
+      if (!kAtlasChars[c]) {
+        c = '_';
+      }
+    }
+    return result;
+  } else {
+    return std::string{input};
```

Comment:
This creates a copy. Maybe change the implementation to take a pointer to the string, and update it in place. Something like:

```c++
#include <algorithm>
// ...
inline void replace_invalid_characters(std::string* s) {
   std::transform(s->begin(), s->end(), s->begin(), [](unsigned char c) {
        return kAtlasChars[c] ? c : '_';
    });
```

Since we are updating the string in place we can avoid the need for `contains_non_atlas_char` since we are not allocating memory in either case.

If that doesn't work, then just `return input;` above instead of `std::string{input}`

Reply:
I kept waffling about whether or not to keep the original Id values intact, so I opted to make the copies here.

Transforming the values in-place should be fine, even if the base Id is affected, because we do not want to allow non-valid characters anyway.